File: 0042-trapping-rain-water/0042-trapping-rain-water.java
import java.util.ArrayDeque;
import java.util.Deque;

class Solution {
    public int trap(int[] height) {
        Deque<Integer> stack = new ArrayDeque<>();
        int water = 0;
        for (int i = 0; i < height.length; i++){

            while (!stack.isEmpty() && height[stack.peek()] < height[i]){
                Integer top = stack.pop();
                if (stack.isEmpty()){
                    break;
                }
                var length = i - stack.peek() - 1;
                int hei = Math.min(height[stack.peek()], height[i]) - height[top];
                water += length * hei;
            }

            stack.push(i);
        }
        return water;
    }
}

File: 0042-trapping-rain-water/0042-trapping-rain-water.kt
class Solution {
    fun trap(height: IntArray): Int {
        var left : Int = 0
        var right : Int = height.size - 1
        var water: Int = 0
        var leftMax : Int = height[left]
        var rightMax : Int = height[right]
        while(left <= right){
            leftMax = maxOf(leftMax, height[left])
            rightMax = maxOf(rightMax, height[right])
            if(leftMax <= rightMax){
                water += leftMax - height[left]
                left++
            }else{
                water += rightMax - height[right]
                right--
            }
        }
        return water
    }
}